scripts/create-darwin-volume.sh

```diff
@@ -0,0 +1,140 @@
+#!/bin/sh
+set -e
+
+root_disk() {
+    diskutil info -plist /
+}
+
+apfs_volumes_for() {
+    disk=$1
+    diskutil apfs list -plist "$disk"
+}
+
+disk_identifier() {
+    xpath "/plist/dict/key[text()='ParentWholeDisk']/following-sibling::string[1]/text()" 2>/dev/null
+}
+
+volume_list_true() {
+    key=$1 t=$2
+    xpath "/plist/dict/array/dict/key[text()='Volumes']/following-sibling::array/dict/key[text()='$key']/following-sibling::true[1]" 2> /dev/null
+}
+
+volume_get_string() {
+    key=$1 i=$2
+    xpath "/plist/dict/array/dict/key[text()='Volumes']/following-sibling::array/dict[$i]/key[text()='$key']/following-sibling::string[1]/text()" 2> /dev/null
+}
+
+find_nix_volume() {
+    disk=$1
+    i=1
+    volumes=$(apfs_volumes_for "$disk")
+    while true; do
+        name=$(echo "$volumes" | volume_get_string "Name" "$i")
+        if [ -z "$name" ]; then
+            break
+        fi
+        case "$name" in
+            [Nn]ix*)
+                echo "$name"
+                break
+                ;;
+        esac
+        i=$((i+1))
+    done
+}
+
+test_fstab() {
+    grep -q "/nix apfs rw" /etc/fstab 2>/dev/null
+}
+
+test_nix_symlink() {
+    [ -L "/nix" ] || grep -q "^nix." /etc/synthetic.conf 2>/dev/null
+}
+
+test_synthetic_conf() {
+    grep -q "^nix$" /etc/synthetic.conf 2>/dev/null
+}
+
+test_nix() {
+    test -d "/nix"
+}
+
+test_filevault() {
+    disk=$1
+    apfs_volumes_for "$disk" | volume_list_true FileVault | grep -q true || return
+    ! sudo xartutil --list >/dev/null 2>/dev/null
+}
+
+main() {
+    (
+        echo ""
+        echo "     ------------------------------------------------------------------ "
+        echo "    | This installer will create a volume for the nix store and        |"
+        echo "    | configure it to mount at /nix.  Follow these steps to uninstall. |"
+        echo "     ------------------------------------------------------------------ "
+        echo ""
+        echo "  1. Remove the entry from fstab using 'sudo vifs'"
+        echo "  2. Destroy the data volume using 'diskutil apfs deleteVolume'"
+        echo "  3. Remove the 'nix' line from /etc/synthetic.conf or the file"
+        echo ""
+    ) >&2
+
+    if test_nix_symlink; then
+        echo "error: /nix is a symlink, please remove it and make sure it's not in synthetic.conf (in which case a reboot is required)" >&2
+        echo "  /nix -> $(readlink "/nix")" >&2
+        exit 2
+    fi
+
+    if ! test_synthetic_conf; then
+        echo "Configuring /etc/synthetic.conf..." >&2
+        echo nix | sudo tee /etc/synthetic.conf
+        if ! test_synthetic_conf; then
+            echo "error: failed to configure synthetic.conf" >&2
+            exit 1
+        fi
+    fi
+
+    if ! test_nix; then
+        echo "Creating mountpoint for /nix..." >&2
+        /System/Library/Filesystems/apfs.fs/Contents/Resources/apfs.util -B || true
```

Comment:
I believe this should instead use

```
apfsctl synthetic folder nix
```

Reply:
Where did you get this from? I don't have that command or a manpage for it on my system.

Reply:
Where does this come from? I don't have `apfsctl` on PATH and `find / -name "apfsctl" 2>/dev/null` came up empty on both Catalina and Mojave. I also only see 8 google results for it. I did find 13 results in a GH code search ([https://github.com/apple-opensource/bless/blob/master/test/test_bless#L544](example)).

Reply:
Also, do you know anything about whether it's possible to get a true firmlink? 😄

Reply:
Sorry, looks like this isn't available. I'll see if I can find another way.

Reply:
By the way, perhaps not exactly intended, but this works fine. The `|| true` is just there as a safeguard in case this ever breaks or moves, etc. Giving the users instructions to reboot rather than failing completely.

Also as mentioned earlier in the thread, symlinks leak the real destination which is undesirable for builds. If it where possible to use a firmlink it would make most of this unnecessary, since I as understand it that would behave similar to a bind mount.

Reply:
If there's any chance at all we could get a true firmlink that would solve a lot of issues. Barring that, if there's any way we could get a reliable solution for ensuring the volume is mounted before the GUI session is restored when encryption is enabled that doesn't involve modifying `/etc/ttys` to set up a login hook or hardcoding the encryption passphrase in a file on disk, that would be really helpful too (ideally it would be part of the same volume group as the system disk, assuming that has anything to do with encryption, but it doesn't seem possible to add new volumes to that group).